heka/io_modules/decoders/heka/json.lua

```diff
@@ -34,18 +48,38 @@ Decode and inject the resulting message
 --]]
 
 -- Imports
+local module_name   = ...
+local module_cfg    = require "string".gsub(module_name, "%.", "_")
 local cjson = require "cjson"
 
 local pairs = pairs
 local type  = type
 
 local inject_message = inject_message
 
+local cfg = read_config(module_cfg) or {}
+assert(type(cfg) == "table", module_cfg .. " must be a table")
+cfg.flatten_delimiter = cfg.flatten_delimiter or "."
+assert(type(cfg.flatten_delimiter) == "string", module_cfg .. ".flatten_delimiter must be a string")
+fields_prefix = "Fields" .. cfg.flatten_delimiter
+
 local M = {}
 setfenv(1, M) -- Remove external access to contain everything in the module
 
 function decode(data, dh)
     local msg = cjson.decode(data)
+    if cfg.preserve_metadata then
```

Comment:
> I don't understand what this comment means as you have access to everything in the current message and the current sandbox environment.

`dh.Type` and `msg.Type` may have different values here.

With the previous behavior there was no way to preserve `dh.Type` so that it could be used in a message matcher on an output.

With the new behavior proposed here, setting `cfg.preserve_metadata` would allow them to be referenced in a message matcher as `Type` and `Fields[Type]` respectively, with the caveat that `msg.Fields.agent` (or similar) would have to be referenced in a message matcher as `Fields[Fields.agent]`